File: 프로그래머스/연습문제/Lv2/연속된 부분 수열의 합/연속된 부분 수열의 합.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> sequence, int k) {
    vector<int> sum;    // 부분합
    sum.push_back(sequence[0]);
    for (int i = 1; i < sequence.size(); i++)
        sum.push_back(sum[i - 1] + sequence[i]);

    int min_dist = sequence.size(), min_idx = 0;
    int start=-1, end=0; // 투 포인터
    for(; sum[end]<k; end++);
    
    while(end < sequence.size()){
        if(start == -1){
            // k값이면 기록
            if(sum[end] == k){
                if(end+1 < min_dist){
                    min_dist = end+1;
                    min_idx = 0;
                }
                end++;
            }
            // k값보다 크면 줄이기
            else if(sum[end] > k) start++;
            // k값보다 크면 늘이기
            else end++;
        }
        else{
            if(sum[end]-sum[start] == k){
                if(end-start < min_dist){
                    min_dist = end-start;
                    min_idx = start+1;
                }
                end++;
            }
            else if(sum[end]-sum[start] > k) start++;
            else end++;
        }
    }

    return { min_idx, min_idx + min_dist - 1 };
}

File: 프로그래머스/연습문제/Lv2/연속된 부분 수열의 합/연속된 부분 수열의 합.py
def solution(sequence, k):
    sums = [sequence[0]]
    for i in range(1, len(sequence)):
        sums.append(sums[i-1]+sequence[i])

    min_dist=len(sequence); min_idx=0
    start,end = -1,0
    while sums[end] < k:
        end += 1

    len_s = len(sequence)
    while end < len_s:
        if start == -1:
            if sums[end] == k:
                if end+1 < min_dist:
                    min_dist = end+1
                    min_idx = 0
                end += 1
            elif sums[end] > k: start += 1
            else: end += 1
        else:
            if sums[end]-sums[start] == k:
                if end-start < min_dist:
                    min_dist = end-start
                    min_idx = start+1
                end += 1
            elif sums[end]-sums[start] > k: start += 1
            else: end += 1

    return [min_idx, min_idx+min_dist-1]
